### PR TITLE
Adding warning for Master as it is deprecated in 5.2

### PR DESCRIPTION
**Pull request by anchuraj**
Since `master` is deprecated from OpenMP spec 5.2, warning is added. Using `masked` is the recommended alternative as per spec

### CODE REVIEW

**Review by tblah — APPROVED** [review]
LGTM, thanks.

I think it would be best to mention which standard version deprecates the directive, but I see that was already not the case for target loop so I won't require that here.

File: flang/lib/Semantics/resolve-directives.cpp [hunk 1]
```diff
@@ -2758,4 +2758,12 @@ void OmpAttributeVisitor::AddOmpRequiresToScope(Scope &scope,
   } while (!scopeIter->IsGlobal());
 }
 
+void OmpAttributeVisitor::IssueNonConformanceWarning(
+    llvm::omp::Directive D, parser::CharBlock source) {
+  if (context_.ShouldWarn(common::UsageWarning::OpenMPUsage)) {
+    context_.Say(source,
+        "Usage of directive %s is non-confirming to OpenMP standard"_warn_en_US,
```
Comment (skatrak) [inline flang/lib/Semantics/resolve-directives.cpp:2765]:
```suggestion
        "Usage of directive '%s' is non-conforming to OpenMP standard"_warn_en_US,
```

**Review by skatrak — COMMENTED** [review]
Implementation LGTM, I'm just wondering if the error message might be a bit too generic. It would be good to let users know that this has been triggered because the construct was deprecated in version 5.2 and that it's recommended to swap it for `masked`.

File: flang/lib/Semantics/resolve-directives.cpp (hunk 1 above)
Comment (mjklemm) [reply flang/lib/Semantics/resolve-directives.cpp:2765]: Maybe something like;

OpenMP directive 'master' has been deprecated; please use 'masked' instead.

Comment (anchuraj) [reply flang/lib/Semantics/resolve-directives.cpp:2765]: Sure.

File: flang/lib/Semantics/resolve-directives.cpp [hunk 2]
```diff
@@ -2758,4 +2758,22 @@ void OmpAttributeVisitor::AddOmpRequiresToScope(Scope &scope,
...
 }
 
+void OmpAttributeVisitor::IssueNonConformanceWarning(
+    llvm::omp::Directive D, parser::CharBlock source) {
+  std::string warnStr = "";
+  std::string dirName = llvm::omp::getOpenMPDirectiveName(D).str();
+  switch (D) {
+  case llvm::omp::OMPD_master:
+    warnStr = "OpenMP directive " + dirName +
+        " has been deprecated, please use masked instead.";
```
Comment (mjklemm) [inline flang/lib/Semantics/resolve-directives.cpp:2768]:
```suggestion
    warnStr = "OpenMP directive '" + dirName +
        "' has been deprecated, please use 'masked' instead.";
```

File: flang/lib/Semantics/resolve-directives.cpp [hunk 3]
```diff
@@ -2758,4 +2758,22 @@ void OmpAttributeVisitor::AddOmpRequiresToScope(Scope &scope,
...
+    break;
+  case llvm::omp::OMPD_target_loop:
+  default:
+    warnStr = "Usage of directive " + dirName +
+        "is non-confirming to OpenMP standard.";
```
Comment (mjklemm) [inline flang/lib/Semantics/resolve-directives.cpp:2773]:
```suggestion
    warnStr = "OpenMP directive '" + dirName +
        "' has been deprecated.";
```

Comment (mjklemm) [reply flang/lib/Semantics/resolve-directives.cpp:2773]: Please only commit this suggestion, if the warning is only hit for deprecated OpenMP directives.

Nit: In any case there's a missing space after the directive name and I would suggest to enclose the directive names in single quotes.

File: flang/test/Semantics/OpenMP/clause-validity01.f90
Comment (mjklemm) [inline flang/test/Semantics/OpenMP/clause-validity01.f90:1]: The tests will have to be updated to match the suggestions for resolve-directives.cpp

**Review by mjklemm — APPROVED** [review]
LGTM when suggestions applied.

File: flang/lib/Semantics/resolve-directives.cpp (hunk 3 above)
Comment (anchuraj) [reply flang/lib/Semantics/resolve-directives.cpp:2773]: Sure. Thank you!

File: flang/test/Semantics/OpenMP/clause-validity01.f90
Comment (anchuraj) [reply flang/test/Semantics/OpenMP/clause-validity01.f90:1]: Thank you for your suggestion :) The latest revision makes the changes suggested.

**Pull request merged**